File: lib/pane.ex
defmodule Pane do
  @moduledoc """
  Pane is a tmux session manager that creates a standardized
  development environment across multiple project directories.
  """
  
  @doc """
  Check if the current process is running in an interactive terminal.
  Uses multiple methods to more reliably detect interactive terminals.
  """
  def is_interactive_terminal? do
    # First approach: Check using test -t 0
    test_tty = 
      case System.cmd("test", ["-t", "0"], stderr_to_stdout: true) do
        {_, 0} -> true
        _ -> false
      end
    
    # Second approach: Check using tty command
    tty_check = 
      case System.cmd("sh", ["-c", "tty | grep -v 'not a tty' > /dev/null"], stderr_to_stdout: true) do
        {_, 0} -> true
        _ -> false
      end
      
    # Third approach: Check standard input as character device
    stdin_check =
      case System.cmd("test", ["-c", "/dev/stdin"], stderr_to_stdout: true) do
        {_, 0} -> true
        _ -> false
      end
    
    # Fourth approach: Check TERM environment variable
    term_check = System.get_env("TERM") && System.get_env("TERM") != ""
    
    # Combine the results - if any method detects a terminal, consider it interactive
    test_tty or tty_check or stdin_check or term_check
  end
  
  @doc """
  Run the tmux session setup with the given configuration.
  
  ## Parameters
    * `config` - The session configuration
    * `auto_attach` - Whether to automatically attach to the session (default: false)
  """
  def run(config, auto_attach \\ false) do
    # Generate all tmux commands
    commands = Pane.Command.generate_commands(config)

    # Check if session exists
    [session_check_cmd | remaining_commands] = commands

    # Last command is the attach command (used for reference only)
    _attach_cmd = List.last(commands)

    session_exists? =
      case System.cmd("sh", ["-c", session_check_cmd], stderr_to_stdout: true) do
        {_, 0} -> true
        _ -> false
      end

    # Only create windows if session doesn't exist
    unless session_exists? do
      # Execute each command except the session check and the attach command
      commands_to_run = Enum.take(remaining_commands, length(remaining_commands) - 1)

      Enum.each(commands_to_run, fn cmd ->
        case System.cmd("sh", ["-c", cmd], stderr_to_stdout: true) do
          {_, 0} ->
            :ok

          {error, _} ->
            IO.puts("Warning: Command failed: #{cmd}")
            IO.puts("Error: #{error}")
        end
      end)
    end

    # Extract the session name from the config
    %{session: session_name} = config

    # Check if we should auto-attach (based on the auto_attach parameter)
    if auto_attach do
      # Check if we're in a terminal environment
      is_interactive = is_interactive_terminal?()
      
      # Log the detection result in verbose mode
      if Application.get_env(:pane, :verbose, false) do
        IO.puts("[INFO] Interactive terminal detection: #{is_interactive}")
      end
      
      if is_interactive do
        # Tell the user what's happening
        IO.puts("\nAttaching to tmux session '#{session_name}'...")
        
        # Prepare the terminal
        IO.write("")
        
        # Use a direct exec approach for attaching to tmux
        exec_cmd = "exec tmux attach-session -t \"#{session_name}\""
        
        # Create a temporary execution script
        temp_script = """
        #!/bin/sh
        # Automatically generated script for tmux attachment
        #{exec_cmd}
        """
        script_path = Path.join(System.tmp_dir(), "pane_attach_#{:os.system_time(:millisecond)}.sh")
        
        # Write and execute the script
        File.write!(script_path, temp_script)
        File.chmod!(script_path, 0o755)
        
        # Execute the script as a process replacement
        Port.open({:spawn_executable, script_path}, [:nouse_stdio])
        :timer.sleep(100) # Small delay to ensure process starts
        
        # In case the exec doesn't work (should be rare), exit gracefully
        System.halt(0)
      else
        # If we're not in an interactive terminal, just print instructions
        IO.puts("\nTmux session '#{session_name}' is ready.")
        IO.puts("Cannot attach automatically - not running in an interactive terminal.")
        IO.puts("To attach to this session, run the following command in your terminal:")
        IO.puts("  tmux attach -t #{session_name}")
      end
    else
      # If auto_attach is false (--no-attach was specified), just print instructions
      IO.puts("\nTmux session '#{session_name}' is ready.")
      IO.puts("To attach to this session, run the following command in your terminal:")
      IO.puts("  tmux attach -t #{session_name}")
      IO.puts("\nRun this script with --preview to see the full list of commands executed.")
    end
  end
end

File: nodejs/bin/pane.js
#!/usr/bin/env node

/**
 * Command-line interface for Pane
 */
const { program } = require('commander');
const { PaneConfig } = require('../src/lib/config');
const { PaneCommand } = require('../src/lib/command');
const { run } = require('../src/index');
const { isInteractiveTerminal } = require('../src/lib/utils');
const { spawn } = require('child_process');
const fs = require('fs');
const tmp = require('tmp');

// Define CLI options
program
  .name('pane')
  .description('A tmux session manager for creating consistent development environments')
  .option('-p, --preview', 'Show commands without executing them')
  .option('-c, --config <config>', 'Use specific config name or file')
  .option('-v, --verbose', 'Show detailed information during execution')
  .option('-a, --attach', 'Directly attach to the session (run this from terminal)')
  .option('--no-attach', 'Create the session but don\'t automatically attach to it')
  .option('--print-session', 'Print the session name from the config and exit')
  .version('0.1.0');

// Parse command line arguments
program.parse(process.argv);

// Get the options
const opts = program.opts();

// Set global flags
if (opts.verbose) {
  global.verbose = true;
}

// Store the no-attach flag for preview mode
global.noAttach = !!opts.noAttach;

// Show help message
function showHelp() {
  console.log(`
Pane - A tmux session manager

Usage:
  pane [options]

Options:
  -p, --preview          Show commands without executing them
  -a, --attach           Directly attach to the session (run this from terminal)
  -c, --config=CONFIG    Use specific config name (e.g. -c myproject) or file (e.g. -c path/to/config.yaml)
  -v, --verbose          Show detailed information during execution
  --no-attach            Create the session but don't automatically attach to it
  --print-session        Print the session name from the config and exit
  -h, --help             Show this help message

Configuration:
  Configs are looked for in the following locations:
    1. As specified (if absolute path)
    2. ~/.config/pane/
    3. The project's config/ directory
  
  Default config (when no -c option) is loaded from:
    1. ~/.config/pane/default.yaml
    2. The project's config/default.yaml file

Examples:
  pane                   Create tmux session using default config and attach to it
  pane -c myproject      Use myproject.yaml config from standard locations
  pane --no-attach       Create session without attaching
  pane --preview         Preview the commands that would be executed
  pane --verbose         Show detailed logging during execution
  pane -v -p             Preview with verbose output
  pane --config=my.yaml  Use custom configuration file
`);
}

// Main process handler
async function process() {
  // Get config file name from command line options
  const configPath = opts.config;
  
  try {
    const config = await PaneConfig.loadConfig(configPath);
    
    if (opts.printSession) {
      console.log(config.session);
      return;
    }
    
    if (opts.preview) {
      // Set preview mode for debug output
      global.previewMode = true;
      
      // Check for no-attach flag
      const noAttach = opts.noAttach === true || opts.attach === false;
      if (noAttach) {
        global.noAttach = true;
      }
      
      PaneCommand.preview(config);
    } else if (opts.attach) {
      // Direct attach mode - execute the attach command directly
      const sessionName = config.session;
      
      // We still log interactive terminal detection for debugging purposes
      const isInteractive = isInteractiveTerminal();
      
      // Log detection result if in verbose mode
      if (global.verbose) {
        console.log('[INFO] Interactive terminal detection: ' + isInteractive);
        console.log('[INFO] Attempting direct attachment');
      }
      
      // Tell the user what's happening
      console.log(`\nAttaching to tmux session '${sessionName}'...`);
      
      // Use a direct exec approach for attaching to tmux
      const execCmd = `exec tmux attach-session -t "${sessionName}"`;
      
      // Create a temporary execution script
      const tempScript = `#!/bin/sh
# Automatically generated script for tmux attachment
${execCmd}
`;
      
      const scriptPath = tmp.tmpNameSync({ prefix: 'pane_attach_', postfix: '.sh' });
      fs.writeFileSync(scriptPath, tempScript);
      fs.chmodSync(scriptPath, 0o755);
      
      // Execute the script as a process replacement
      spawn(scriptPath, [], {
        stdio: 'inherit',
        detached: true
      });
      
      // Small delay to ensure process starts
      setTimeout(() => {
        process.exit(0);
      }, 100);
    } else {
      // Run in normal mode without preview
      global.previewMode = false;
      
      // Only auto-attach if --no-attach isn't specified
      const noAttach = opts.noAttach === true || opts.attach === false;
      const autoAttach = !noAttach;
      
      // Log the auto-attach setting in verbose mode
      if (global.verbose) {
        console.log(`[INFO] Auto-attach enabled: ${autoAttach}`);
      }
      
      run(config, autoAttach);
    }
  } catch (error) {
    console.error(`Error: Could not load configuration: ${error.message}`);
    process.exit(1);
  }
}

// Handle program options
if (program.args.length === 0 && Object.keys(opts).length === 0) {
  // No arguments or options - show help
  showHelp();
} else if (opts.help) {
  // Help option - show help
  showHelp();
} else {
  // Process other options
  process();
}